tests/structured_conf/test_structured_config.py

```diff
@@ -1051,3 +1047,130 @@ def test_update_userlist(
+        cfg = OmegaConf.structured(module.OptionalBook)
+        cfg.author = None
+
+    def test_non_optional_union_set_none(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.Book)
+        with pytest.raises(ValidationError):
+            cfg.author = None
+
+    @pytest.mark.parametrize(  # type: ignore
+        "obj,value,node",
+        [
+            ("DictUnion", {"foo": 0.33, "foo2": 10}, "dict"),
+            ("ListUnion", [1, 2, 0.33], "list"),
+        ],
+    )
+    def test_cfg_set_valid_union_value(
+        self, class_type: str, obj: str, value: Any, node: str
+    ) -> None:
+        module: Any = import_module(class_type)
+        class_ = getattr(module, obj)
+        cfg = OmegaConf.structured(class_)
+        cfg[node] = value
+        assert cfg[node] == value
+
+    @pytest.mark.parametrize(  # type: ignore
+        "obj,value,node",
+        [
+            ("DictUnion", {"foo": False}, "dict"),
+            ("DictUnion", {"foo": "invalid"}, "dict"),
+            ("DictUnion", {"foo": Color.BLUE}, "dict"),
+            ("DictUnion", {"foo": User()}, "dict"),
+            ("DictUnion", {"foo": None}, "dict"),
+            ("ListUnion", [1, 2, False], "list"),
+            ("ListUnion", [1, 2, "invalid"], "list"),
+            ("ListUnion", [1, 2, Color.BLUE], "list"),
+            ("ListUnion", [1, 2, User()], "list"),
+            ("ListUnion", [1, 2, None], "list"),
+        ],
+    )
+    def test_cfg_set_invalid_union_value(
+        self, class_type: str, obj: str, value: Any, node: str
+    ) -> None:
+        module: Any = import_module(class_type)
+        class_ = getattr(module, obj)
+        cfg = OmegaConf.structured(class_)
+        with pytest.raises(ValidationError):
+            cfg[node] = value
+
+    @pytest.mark.parametrize(  # type: ignore
+        "value,node",
+        [
+            (DictConfig({"foo": False}), "union_dict"),
+            (ListConfig([1, 2, False]), "union_list"),
+        ],
+    )
+    def test_cfg_set_container_in_union(
+        self, class_type: str, node: str, value: Any
+    ) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.UnionWithContainer)
+        cfg[node] = value
+        assert cfg[node] == value
+        assert isinstance(cfg._get_node(node), UnionNode)
+
+    def test_union_with_baseclass(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.UnionWithBaseclass)
+        cfg.foo = module.Subclass1()
+        assert cfg.foo == module.Subclass1()
+        assert isinstance(cfg._get_node("foo"), UnionNode)
+
+    def test_union_with_subclasses(self, class_type: str) -> None:
+        module: Any = import_module(class_type)
+        cfg = OmegaConf.structured(module.UnionOfSubclasses)
+        cfg.foo = module.Subclass1
+        assert cfg.foo == module.Subclass1()
```

Comment:
I'm a bit surprised this works, since `cfg.foo` is set to the class and not an instance. What's going on there?